sim/hunter/talents.go

```diff
@@ -1,870 +1,823 @@
-	// }
-
-	// hunter.applySpiritBond()
-	// hunter.applyInvigoration()
-	// hunter.applyCobraStrikes()
-	// hunter.applyGoForTheThroat()
-	// hunter.applyPiercingShots()
-	// hunter.applyWildQuiver()
-	// hunter.applyImprovedTracking()
-	// hunter.applyThrillOfTheHunt()
-	// hunter.applyLockAndLoad()
-	// hunter.applyExposeWeakness()
-	// hunter.applyMasterTactician()
-	// hunter.applySniperTraining()
-	// hunter.applyHuntingParty()
-
-	// hunter.registerReadinessCD()
+	if hunter.Pet != nil {
+		hunter.applyFrenzy()
+		hunter.registerBestialWrathCD()
+		// Todo: BM stuff
+		hunter.Pet.ApplyTalents()
+	}
+
+	if hunter.Talents.Pathing > 0 {
+		bonus := 0.03 * float64(hunter.Talents.Pathing)
+		hunter.MultiplyCastSpeed(bonus)
+	}
+
+	if hunter.Talents.HunterVsWild > 0 {
+		bonus := 0.05 * float64(hunter.Talents.HunterVsWild)
+		hunter.MultiplyStat(stats.Stamina, 1+bonus)
+	}
+
+	if hunter.Talents.HuntingParty {
+		agiBonus := 0.02
+		hunter.MultiplyStat(stats.Agility, 1.0+agiBonus)
+	}
+
+	if hunter.Talents.KillingStreak > 0 {
+		hunter.applyKillingStreak()
+	}
+
+	hunter.applyCobraStrikes()
+	hunter.applyPiercingShots()
+	hunter.applySpiritBond()
+	hunter.applyInvigoration()
+	hunter.applyGoForTheThroat()
+	hunter.applyThrillOfTheHunt()
+	hunter.applyTNT()
+	hunter.applySniperTraining()
+	hunter.applyHuntingParty()
+	hunter.applyImprovedSteadyShot()
+	hunter.applyFocusFireCD()
+	hunter.applyFervorCD()
+	hunter.registerReadinessCD()
+	hunter.applyMasterMarksman()
+	hunter.applyTermination()
+}
+func (hunter *Hunter) applyMasterMarksman() {
+	if hunter.Talents.MasterMarksman == 0 {
+		return
+	}
+
+	procChance := float64(hunter.Talents.MasterMarksman) * 0.2
+	hunter.MasterMarksmanAura = hunter.RegisterAura(core.Aura{
+		Label:    "Ready, Set, Aim...",
+		ActionID: core.ActionID{SpellID: 82925},
+		Duration: time.Second * 8,
+		OnGain: func(aura *core.Aura, sim *core.Simulation) {
+			if hunter.AimedShot != nil {
+				hunter.AimedShot.CostMultiplier = 0
+				hunter.AimedShot.DefaultCast.CastTime = 0
+			}
+		},
+		OnExpire: func(aura *core.Aura, sim *core.Simulation) {
+			if hunter.AimedShot != nil {
+				hunter.AimedShot.CostMultiplier = 1
+				hunter.AimedShot.DefaultCast.CastTime = time.Second * 3
+			}
+		},
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if spell == hunter.AimedShot {
+				hunter.MasterMarksmanCounterAura.SetStacks(sim, 0)
+				hunter.MasterMarksmanCounterAura.Activate(sim)
+				aura.Deactivate(sim) // Consume effect
+			}
+
+		},
+	})
+	hunter.MasterMarksmanCounterAura = hunter.RegisterAura(core.Aura{
+		Label:     "Master Marksman",
+		Duration:  time.Second * 30,
+		ActionID:  core.ActionID{SpellID: 34486},
+		MaxStacks: 4,
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if spell != hunter.SteadyShot {
+				return
+			}
+			if sim.Proc(procChance, "Master Marksman Proc") {
```

Comment:
you can optimize this check by checking `if procChance == 1 || sim.Proc(...` this would compltely skip rolling the random as you have 100% chance to apply it anyway